### ISSUE TITLE
Race in prometheus exporter

### ISSUE THREAD

**Issue by rakyll** [census-instrumentation/opencensus-go#402]
See the log:

```
fatal error: concurrent map iteration and map write

goroutine 198 [running]:
runtime.throw(0x1587a35, 0x26)
	/Users/jbd/go/src/runtime/panic.go:619 +0x81 fp=0xc4205d4e08 sp=0xc4205d4de8 pc=0x102be41
runtime.mapiternext(0xc4205d4f28)
	/Users/jbd/go/src/runtime/hashmap.go:747 +0x55c fp=0xc4205d4e98 sp=0xc4205d4e08 pc=0x100aacc
go.opencensus.io/exporter/prometheus.(*collector).Collect(0xc4201a86c0, 0xc420324120)
	/Users/jbd/src/go.opencensus.io/exporter/prometheus/prometheus.go:206 +0x92 fp=0xc4205d4f98 sp=0xc4205d4e98 pc=0x13f1db2
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func2(0xc4203ec0d0, 0xc420324120, 0x15da7e0, 0xc4201a86c0)
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/registry.go:382 +0x61 fp=0xc4205d4fc0 sp=0xc4205d4f98 pc=0x13ef191
runtime.goexit()
	/Users/jbd/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4205d4fc8 sp=0xc4205d4fc0 pc=0x10573a1
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/registry.go:380 +0x302

goroutine 1 [sleep]:
time.Sleep(0x5f5e100)
	/Users/jbd/go/src/runtime/time.go:102 +0x166
main.main()
	/Users/jbd/src/github.com/rakyll/opencensus-grpc-demo/client/main.go:81 +0x598

goroutine 20 [select]:
go.opencensus.io/stats.(*worker).start(0xc4200ae180)
	/Users/jbd/src/go.opencensus.io/stats/worker.go:221 +0x11d
created by go.opencensus.io/stats.init.0
	/Users/jbd/src/go.opencensus.io/stats/worker.go:29 +0x54

goroutine 38 [select]:
google.golang.org/grpc.(*ccResolverWrapper).watcher(0xc4201f5ad0)
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:110 +0x182
created by google.golang.org/grpc.(*ccResolverWrapper).start
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:96 +0x3f

goroutine 5 [select]:
google.golang.org/grpc.(*addrConn).transportMonitor(0xc420254000)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:1238 +0x235
google.golang.org/grpc.(*addrConn).connect.func1(0xc420254000)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:841 +0x1b5
created by google.golang.org/grpc.(*addrConn).connect
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:832 +0xe1

goroutine 4 [select]:
google.golang.org/grpc.(*ccBalancerWrapper).watcher(0xc42002e480)
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:122 +0x14a
created by google.golang.org/grpc.newCCBalancerWrapper
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:113 +0x14c

goroutine 24 [select]:
google.golang.org/grpc.(*ccBalancerWrapper).watcher(0xc4200ae600)
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:122 +0x14a
created by google.golang.org/grpc.newCCBalancerWrapper
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:113 +0x14c

goroutine 39 [select]:
google.golang.org/grpc.(*ccResolverWrapper).watcher(0xc4201f5c50)
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:110 +0x182
created by google.golang.org/grpc.(*ccResolverWrapper).start
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:96 +0x3f

goroutine 40 [IO wait]:
internal/poll.runtime_pollWait(0x1d50ea0, 0x72, 0x0)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4202fc018, 0x72, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4202fc018, 0xffffffffffffff00, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc4202fc000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:372 +0x1a8
net.(*netFD).accept(0xc4202fc000, 0x14b04e0, 0xc4202e0de8, 0x1003418)
	/Users/jbd/go/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc4202fe000, 0xc4202e0e18, 0x1002377, 0xc420502800)
	/Users/jbd/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).AcceptTCP(0xc4202fe000, 0xc4202e0e60, 0xc4202e0e68, 0x18)
	/Users/jbd/go/src/net/tcpsock.go:246 +0x49
net/http.tcpKeepAliveListener.Accept(0xc4202fe000, 0x159bf20, 0xc420502780, 0x15dffe0, 0xc4202f20f0)
	/Users/jbd/go/src/net/http/server.go:3216 +0x2f
net/http.(*Server).Serve(0xc42028c000, 0x15dfd60, 0xc4202fe000, 0x0, 0x0)
	/Users/jbd/go/src/net/http/server.go:2770 +0x1a5
net/http.(*Server).ListenAndServe(0xc42028c000, 0xc42028c000, 0x159b708)
	/Users/jbd/go/src/net/http/server.go:2711 +0xa9
net/http.ListenAndServe(0x1573772, 0x5, 0x0, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/http/server.go:2969 +0x7a
main.main.func1(0xc4201a8740)
	/Users/jbd/src/github.com/rakyll/opencensus-grpc-demo/client/main.go:52 +0x76
created by main.main
	/Users/jbd/src/github.com/rakyll/opencensus-grpc-demo/client/main.go:47 +0x26e

goroutine 25 [select]:
google.golang.org/grpc.(*addrConn).transportMonitor(0xc4200cef00)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:1238 +0x235
google.golang.org/grpc.(*addrConn).connect.func1(0xc4200cef00)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:841 +0x1b5
created by google.golang.org/grpc.(*addrConn).connect
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:832 +0xe1

goroutine 42 [select]:
google.golang.org/grpc.(*ccResolverWrapper).watcher(0xc4201f5dd0)
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:110 +0x182
created by google.golang.org/grpc.(*ccResolverWrapper).start
	/Users/jbd/src/google.golang.org/grpc/resolver_conn_wrapper.go:96 +0x3f

goroutine 50 [select]:
google.golang.org/grpc.(*ccBalancerWrapper).watcher(0xc4202a00c0)
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:122 +0x14a
created by google.golang.org/grpc.newCCBalancerWrapper
	/Users/jbd/src/google.golang.org/grpc/balancer_conn_wrappers.go:113 +0x14c

goroutine 51 [select]:
google.golang.org/grpc.(*addrConn).createTransport(0xc4202c2000, 0x6, 0x0, 0xbe95dbdba9b778f1, 0xa66a073bd, 0x1873620, 0xbe95dbdbe8485ad3, 0xaa0cc1f9f, 0x1873620, 0xc4200aeac0, ...)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:1215 +0xd81
google.golang.org/grpc.(*addrConn).resetTransport(0xc4202c2000, 0x0, 0xc42029e060)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:1104 +0x4e5
google.golang.org/grpc.(*addrConn).connect.func1(0xc4202c2000)
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:833 +0x2f
created by google.golang.org/grpc.(*addrConn).connect
	/Users/jbd/src/google.golang.org/grpc/clientconn.go:832 +0xe1

goroutine 67 [IO wait]:
internal/poll.runtime_pollWait(0x1d50dd0, 0x72, 0xc4202de970)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4201de298, 0x72, 0xffffffffffffff00, 0x15d9940, 0x18285f8)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4201de298, 0xc42033c000, 0x1000, 0x1000)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4201de280, 0xc42033c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:157 +0x1dc
net.(*netFD).Read(0xc4201de280, 0xc42033c000, 0x1000, 0x1000, 0x8, 0x8, 0xff3)
	/Users/jbd/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc42000e138, 0xc42033c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/net.go:176 +0x6a
crypto/tls.(*block).readFromUntil(0xc4201e98f0, 0x31a00a0, 0xc42000e138, 0x5, 0xc42000e138, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:493 +0x96
crypto/tls.(*Conn).readRecord(0xc420334000, 0x159c517, 0xc420334120, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:595 +0xe0
crypto/tls.(*Conn).Read(0xc420334000, 0xc4205bc000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:1156 +0x100
bufio.(*Reader).Read(0xc4203242a0, 0xc4205cc038, 0x9, 0x9, 0x30, 0x28, 0x151d6c0)
	/Users/jbd/go/src/bufio/bufio.go:216 +0x238
io.ReadAtLeast(0x15d7760, 0xc4203242a0, 0xc4205cc038, 0x9, 0x9, 0x9, 0xc4202dedf0, 0x10024ae, 0xc4202dee9f)
	/Users/jbd/go/src/io/io.go:309 +0x86
io.ReadFull(0x15d7760, 0xc4203242a0, 0xc4205cc038, 0x9, 0x9, 0x1322964, 0xc4202e42a0, 0xc4202e0004)
	/Users/jbd/go/src/io/io.go:327 +0x58
golang.org/x/net/http2.readFrameHeader(0xc4205cc038, 0x9, 0x9, 0x15d7760, 0xc4203242a0, 0x0, 0x0, 0xc4202e42a0, 0x0)
	/Users/jbd/src/golang.org/x/net/http2/frame.go:237 +0x7b
golang.org/x/net/http2.(*Framer).ReadFrame(0xc4205cc000, 0xc4202e42a0, 0xc4202e4200, 0x0, 0x0)
	/Users/jbd/src/golang.org/x/net/http2/frame.go:492 +0xa4
google.golang.org/grpc/transport.(*http2Client).reader(0xc4202b0600)
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:1177 +0xe1
created by google.golang.org/grpc/transport.newHTTP2Client
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:273 +0xbf1

goroutine 16 [select]:
google.golang.org/grpc/transport.loopyWriter(0x15dff20, 0xc42002e540, 0xc4205a2120, 0xc4202e2fb8)
	/Users/jbd/src/google.golang.org/grpc/transport/transport.go:727 +0x368
google.golang.org/grpc/transport.newHTTP2Client.func3(0xc420280d80)
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:305 +0x5e
created by google.golang.org/grpc/transport.newHTTP2Client
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:304 +0xd42

goroutine 30 [IO wait]:
internal/poll.runtime_pollWait(0x1d50d00, 0x72, 0xc420295970)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4202da198, 0x72, 0xffffffffffffff00, 0x15d9940, 0x18285f8)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4202da198, 0xc420362000, 0x1000, 0x1000)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4202da180, 0xc420362000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:157 +0x1dc
net.(*netFD).Read(0xc4202da180, 0xc420362000, 0x1000, 0x1000, 0x8, 0x8, 0xff3)
	/Users/jbd/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc4200ba120, 0xc420362000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/net.go:176 +0x6a
crypto/tls.(*block).readFromUntil(0xc420193b30, 0x31a00a0, 0xc4200ba120, 0x5, 0xc4200ba120, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:493 +0x96
crypto/tls.(*Conn).readRecord(0xc4200b3880, 0x159c517, 0xc4200b39a0, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:595 +0xe0
crypto/tls.(*Conn).Read(0xc4200b3880, 0xc4205a4000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:1156 +0x100
bufio.(*Reader).Read(0xc42019f140, 0xc4205b4038, 0x9, 0x9, 0x20, 0x0, 0x0)
	/Users/jbd/go/src/bufio/bufio.go:216 +0x238
io.ReadAtLeast(0x15d7760, 0xc42019f140, 0xc4205b4038, 0x9, 0x9, 0x9, 0xc420295df0, 0x10024ae, 0xc420295e9f)
	/Users/jbd/go/src/io/io.go:309 +0x86
io.ReadFull(0x15d7760, 0xc42019f140, 0xc4205b4038, 0x9, 0x9, 0x132312d, 0xc4200cae6c, 0xc420574420)
	/Users/jbd/go/src/io/io.go:327 +0x58
golang.org/x/net/http2.readFrameHeader(0xc4205b4038, 0x9, 0x9, 0x15d7760, 0xc42019f140, 0x0, 0xbe95dbd800000000, 0x7b7d7d6e5, 0x1873620)
	/Users/jbd/src/golang.org/x/net/http2/frame.go:237 +0x7b
golang.org/x/net/http2.(*Framer).ReadFrame(0xc4205b4000, 0xc4200cae60, 0xc4200cae60, 0x0, 0x0)
	/Users/jbd/src/golang.org/x/net/http2/frame.go:492 +0xa4
google.golang.org/grpc/transport.(*http2Client).reader(0xc420280d80)
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:1177 +0xe1
created by google.golang.org/grpc/transport.newHTTP2Client
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:273 +0xbf1

goroutine 31 [select]:
google.golang.org/grpc/transport.loopyWriter(0x15dff20, 0xc4200ae6c0, 0xc42052c2d0, 0xc4202dcfb8)
	/Users/jbd/src/google.golang.org/grpc/transport/transport.go:727 +0x368
google.golang.org/grpc/transport.newHTTP2Client.func3(0xc4202b0600)
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:305 +0x5e
created by google.golang.org/grpc/transport.newHTTP2Client
	/Users/jbd/src/google.golang.org/grpc/transport/http2_client.go:304 +0xd42

goroutine 197 [semacquire]:
sync.runtime_Semacquire(0xc4203ec0dc)
	/Users/jbd/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc4203ec0d0)
	/Users/jbd/go/src/sync/waitgroup.go:129 +0x72
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1(0xc4203ec0d0, 0xc420324120)
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/registry.go:376 +0x2b
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/registry.go:375 +0x265

goroutine 196 [IO wait]:
internal/poll.runtime_pollWait(0x1d50b60, 0x72, 0xc4205d1658)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4201c5218, 0x72, 0xffffffffffffff00, 0x15d9940, 0x18285f8)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4201c5218, 0xc42052de00, 0x1, 0x1)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4201c5200, 0xc42052de11, 0x1, 0x1, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:157 +0x1dc
net.(*netFD).Read(0xc4201c5200, 0xc42052de11, 0x1, 0x1, 0xc42008e540, 0x187f2c0, 0x187f2d0)
	/Users/jbd/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc4202fe458, 0xc42052de11, 0x1, 0x1, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/net.go:176 +0x6a
net/http.(*connReader).backgroundRead(0xc42052de00)
	/Users/jbd/go/src/net/http/server.go:668 +0x5a
created by net/http.(*connReader).startBackgroundRead
	/Users/jbd/go/src/net/http/server.go:664 +0xce

goroutine 115 [IO wait]:
internal/poll.runtime_pollWait(0x1d50c30, 0x72, 0xc420299858)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4202daf98, 0x72, 0xffffffffffffff00, 0x15d9940, 0x18285f8)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4202daf98, 0xc42030f000, 0x800, 0x800)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4202daf80, 0xc42030f000, 0x800, 0x800, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:157 +0x1dc
net.(*netFD).Read(0xc4202daf80, 0xc42030f000, 0x800, 0x800, 0x1034a99, 0xc420281b00, 0x200000003)
	/Users/jbd/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc4202fe278, 0xc42030f000, 0x800, 0x800, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/net.go:176 +0x6a
crypto/tls.(*block).readFromUntil(0xc4202f3890, 0x31a00a0, 0xc4202fe278, 0x5, 0xc4202fe278, 0xc420299a10)
	/Users/jbd/go/src/crypto/tls/conn.go:493 +0x96
crypto/tls.(*Conn).readRecord(0xc4200b2380, 0x159c517, 0xc4200b24a0, 0xc420281b00)
	/Users/jbd/go/src/crypto/tls/conn.go:595 +0xe0
crypto/tls.(*Conn).Read(0xc4200b2380, 0xc42029a000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/crypto/tls/conn.go:1156 +0x100
bufio.(*Reader).Read(0xc420324360, 0xc4201a48f8, 0x9, 0x9, 0xc4200b26b0, 0x11d2390, 0xc4200b26b0)
	/Users/jbd/go/src/bufio/bufio.go:216 +0x238
io.ReadAtLeast(0x15d7760, 0xc420324360, 0xc4201a48f8, 0x9, 0x9, 0x9, 0x124a1e1, 0xc4200b2380, 0xc4203d1000)
	/Users/jbd/go/src/io/io.go:309 +0x86
io.ReadFull(0x15d7760, 0xc420324360, 0xc4201a48f8, 0x9, 0x9, 0x1290b64, 0x1d1c3d8, 0xc4200b2380)
	/Users/jbd/go/src/io/io.go:327 +0x58
net/http.http2readFrameHeader(0xc4201a48f8, 0x9, 0x9, 0x15d7760, 0xc420324360, 0x0, 0xc400000000, 0x102b254, 0x159c5d8)
	/Users/jbd/go/src/net/http/h2_bundle.go:1517 +0x7b
net/http.(*http2Framer).ReadFrame(0xc4201a48c0, 0xc4204ea2a0, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/http/h2_bundle.go:1775 +0xa4
net/http.(*http2clientConnReadLoop).run(0xc420299fb8, 0x159bf68, 0xc4201bd7b8)
	/Users/jbd/go/src/net/http/h2_bundle.go:8028 +0x90
net/http.(*http2ClientConn).readLoop(0xc420356000)
	/Users/jbd/go/src/net/http/h2_bundle.go:7956 +0x68
created by net/http.(*http2Transport).newClientConn
	/Users/jbd/go/src/net/http/h2_bundle.go:7143 +0x65c

goroutine 148 [chan receive]:
github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0xc4201a8680, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/registry.go:404 +0x4ed
github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1(0x15df8e0, 0xc4203167e0, 0xc4204d2100)
	/Users/jbd/src/github.com/prometheus/client_golang/prometheus/promhttp/http.go:82 +0x49
net/http.HandlerFunc.ServeHTTP(0xc4201a8700, 0x15df8e0, 0xc4203167e0, 0xc4204d2100)
	/Users/jbd/go/src/net/http/server.go:1947 +0x44
go.opencensus.io/exporter/prometheus.(*Exporter).ServeHTTP(0xc4201a8740, 0x15df8e0, 0xc4203167e0, 0xc4204d2100)
	/Users/jbd/src/go.opencensus.io/exporter/prometheus/prometheus.go:157 +0x52
net/http.(*ServeMux).ServeHTTP(0x1873060, 0x15df8e0, 0xc4203167e0, 0xc4204d2100)
	/Users/jbd/go/src/net/http/server.go:2337 +0x130
net/http.serverHandler.ServeHTTP(0xc42028c000, 0x15df8e0, 0xc4203167e0, 0xc4204d2100)
	/Users/jbd/go/src/net/http/server.go:2694 +0xbc
net/http.(*conn).serve(0xc420502d20, 0x15dff20, 0xc4201a8d80)
	/Users/jbd/go/src/net/http/server.go:1830 +0x651
created by net/http.(*Server).Serve
	/Users/jbd/go/src/net/http/server.go:2795 +0x27b

goroutine 218 [IO wait]:
internal/poll.runtime_pollWait(0x1d50a90, 0x72, 0xc4202e1868)
	/Users/jbd/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4201c5118, 0x72, 0xffffffffffffff00, 0x15d9940, 0x18285f8)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4201c5118, 0xc420360000, 0x1000, 0x1000)
	/Users/jbd/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4201c5100, 0xc420360000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/internal/poll/fd_unix.go:157 +0x1dc
net.(*netFD).Read(0xc4201c5100, 0xc420360000, 0x1000, 0x1000, 0xc4202e1978, 0x1263ab6, 0x1520f00)
	/Users/jbd/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc4202be3d0, 0xc420360000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/net.go:176 +0x6a
net/http.(*connReader).Read(0xc420383770, 0xc420360000, 0x1000, 0x1000, 0xc4202e1a01, 0x100000101347c, 0x3)
	/Users/jbd/go/src/net/http/server.go:764 +0xf8
bufio.(*Reader).fill(0xc4202bb2c0)
	/Users/jbd/go/src/bufio/bufio.go:100 +0x11e
bufio.(*Reader).ReadSlice(0xc4202bb2c0, 0xc42025d00a, 0xc4202e1ab0, 0x101fc76, 0xc4203837a0, 0x1de1ba7, 0xc4202e1a40)
	/Users/jbd/go/src/bufio/bufio.go:341 +0x2c
bufio.(*Reader).ReadLine(0xc4202bb2c0, 0xc4204d2a00, 0x100, 0xf8, 0x1564ac0, 0x0, 0xc420055000)
	/Users/jbd/go/src/bufio/bufio.go:370 +0x34
net/textproto.(*Reader).readLineSlice(0xc4203837a0, 0x57, 0xc4202e1ba8, 0x1012998, 0x100, 0x1564ac0)
	/Users/jbd/go/src/net/textproto/reader.go:55 +0x70
net/textproto.(*Reader).ReadLine(0xc4203837a0, 0xc4204d2a00, 0x0, 0x0, 0x72)
	/Users/jbd/go/src/net/textproto/reader.go:36 +0x2b
net/http.readRequest(0xc4202bb2c0, 0x0, 0xc4204d2a00, 0x0, 0x0)
	/Users/jbd/go/src/net/http/request.go:929 +0x8c
net/http.(*conn).readRequest(0xc420502780, 0x15dff20, 0xc4205dc600, 0x0, 0x0, 0x0)
	/Users/jbd/go/src/net/http/server.go:944 +0x16f
net/http.(*conn).serve(0xc420502780, 0x15dff20, 0xc4205dc600)
	/Users/jbd/go/src/net/http/server.go:1768 +0x4dc
created by net/http.(*Server).Serve
	/Users/jbd/go/src/net/http/server.go:2795 +0x27b
exit status 2
```

**Comment by odeke-em** [comment]
Thank you @rakyll for the report. Here is a test reproducing it:

```go
func TestCollectNonRacy(t *testing.T) {
	exp, err := NewExporter(Options{})
	if err != nil {
		t.Fatalf("NewExporter: %v", err)
	}
	collector := exp.c

	// Synchronize and make sure every goroutine has terminated before we exit
	var waiter sync.WaitGroup
	waiter.Add(3)
	defer waiter.Wait()

	doneCh := make(chan bool)
	// 1. Viewdata write routine at 10ms
	go func() {
		defer waiter.Done()
		tick := time.NewTicker(10 * time.Millisecond)
		defer tick.Stop()

		for {
			select {
			case <-doneCh:
				return
			case <-tick.C:
				count1 := stats.CountData(1)
				mean1 := &stats.MeanData{Mean: 4.5, Count: 5}
				vds := []*stats.ViewData{
					{View: newView(stats.MeanAggregation{}, stats.Cumulative{}), Rows: []*stats.Row{{nil, mean1}}},
					{View: newView(stats.CountAggregation{}, stats.Cumulative{}), Rows: []*stats.Row{{nil, &count1}}},
				}
				for _, v := range vds {
					exp.ExportView(v)
				}
			}
		}
	}()

	inMetricsChan := make(chan prometheus.Metric, 1000)
	// 2. Simulating the Prometheus metrics consumption routine at 11ms
	go func() {
		defer waiter.Done()
		tick := time.NewTicker(11 * time.Millisecond)
		defer tick.Stop()

		n := int(1e5)
		for i := 0; i < n; i++ {
			select {
			case <-doneCh:
				return
			case <-inMetricsChan:
			}
		}
	}()

	// 3. Collect/Read routine at 9ms
	go func() {
		defer waiter.Done()
		tick := time.NewTicker(9 * time.Millisecond)
		defer tick.Stop()

		for {
			select {
			case <-doneCh:
				return
			case <-tick.C:
				// Perform some collection here
				collector.Collect(inMetricsChan)
			}
		}
	}()
}
```

The oversight comes from the assumption 
https://github.com/census-instrumentation/opencensus-go/blob/7a3c63ab1c5772f940fb091832581522533e432f/exporter/prometheus/prometheus.go#L205-L206
We need to create a copy of the views that we have at that moment that Collect is invoked.

The rest of the code in that loop looks well protected against concurrent access.

**Comment by odeke-em** [comment]
Thank you @rakyll for reporting this issue. It has been fixed by #405, please get the latest from master.
